lightning/src/ln/channelmanager.rs

```diff
@@ -760,12 +768,36 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 			pending_events: Mutex::new(Vec::new()),
 			total_consistency_lock: RwLock::new(()),
 
+			persistence_lock: Arc::new((Mutex::new(false), Condvar::new())),
+
 			keys_manager,
 
 			logger,
 		}
 	}
 
+	/// Used to signal to users when the ChannelManager needs to be persisted to disk. This function
+	/// blocks either for the given Duration or indefinitely until the ChannelManager has updates and
+	/// needs persistence. It returns a bool indicating whether persistence is necessary (which will
+	/// always be true if max_wait is None).
+	pub fn block_until_needs_persist(&self, max_wait: Option<Duration>) -> bool {
+		loop {
+			let mutcond = Arc::clone(&self.persistence_lock);
+			let &(ref mtx, ref cvar) = &*mutcond;
+			let mut guard = mtx.lock().unwrap();
+			let result = match max_wait {
+				Some(wait) => cvar.wait_timeout(guard, wait).unwrap().0,
```

Comment:
The doc for `wait_timeout` recommends reducing the wait amount by the actual time elapsed, so that spurious wakeups don't end up making you wait for a multiple of the wait amount as you go around this loop.

Reply:
Alternatively, building this on top of `mpsc::channel` might be less complex / low-level.

Reply:
That seems like a lot of extra overhead just to avoid hand-writing a condition variable wait pattern. I think we basically just copy the code out of the documentation for condition variables here and call it a day - its such an incredibly common pattern I'm not really sure why we should take the limitations of channels here.

Reply:
Fixed the spurious wakeups issue, but stuck with the condvar.